contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,144 @@
+    structure: DataStruct,
+    /// The type inside the structure: struct MyDb(ThisType).
+    connection_type: Type,
+}
+
+const EXAMPLE: &str = "example: `struct MyDatabase(diesel::SqliteConnection);`";
+const ONLY_ON_STRUCTS_MSG: &str = "`database` attribute can only be used on structs";
+const ONLY_UNNAMED_FIELDS: &str = "`database` attribute can only be applied to structs with \
+    exactly one unnamed field";
+const NO_GENERIC_STRUCTS: &str = "`database` attribute cannot be applied to a struct with a \
+    generic type";
+
+const UNABLE_TO_APPLY: &str = "unable to apply the `database` attribute to the \
+    given struct";
+
+fn parse_invocation(attr: TokenStream, input: TokenStream) -> Result<DatabaseInvocation> {
+    let attr_stream2 = ::proc_macro2::TokenStream::from(attr);
+    let attr_span = attr_stream2.span();
+    let string_lit = ::syn::parse2::<LitStr>(attr_stream2)
+        .map_err(|_| attr_span.error("expected string literal"))?;
+
+    let input = ::syn::parse::<DeriveInput>(input).unwrap();
+
+    if !input.generics.params.is_empty() {
+        return Err(input.span().error(NO_GENERIC_STRUCTS));
+    }
+
+    let structure = match input.data {
+        Data::Struct(s) => s,
+        _ => return Err(input.span().error(ONLY_ON_STRUCTS_MSG))
+    };
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    if let Type::Path(type_path) = invocation.connection_type {
+        let inner_type = &type_path.path.segments.iter()
+            .map(|ps| ps.ident.clone())
+            .collect::<Vec<Ident>>();
+
+        let database_name = &invocation.db_name;
+        let handler_type = &invocation.type_name;
+        let pool_type = Ident::new(&format!("{}Pool", handler_type), Span::call_site());
+
+        let tokens = quote! {
+            use ::std::ops::*;
+            use ::rocket::http::*;
+            use ::rocket::request::*;
+            use ::rocket::*;
+            use ::rocket::fairing::*;
+            use ::rocket_contrib::*;
+            use ::rocket_contrib::databases::*;
+            pub struct #handler_type(r2d2::PooledConnection<<#(#inner_type)::* as Poolable>::Manager>);
+            pub struct #pool_type(r2d2::Pool<<#(#inner_type)::* as Poolable>::Manager>);
+
+            impl #handler_type {
+                pub fn pool(config: DatabaseConfig) -> Result<#pool_type, <#(#inner_type)::* as Poolable>::Error> {
```

Comment:
I think we can improve the implementation of this method. Here's one (unchecked) idea:

```rust
    AdHoc::on_attach(|rocket| {
        databases::database_config(#database_name, &rocket.config())
            .map(Self::pool)
            .map(rocket.manage)
            .map_err(|rocket| {
                error_!("Could not find database: {}", #database_name);
                Err(rocket)
            })
    })
```

Reply:
I think the example is good, but one implementation detail we'd need to work out is how to return `Rocket` from the `database_config` method. Otherwise, this wouldn't work inside the `AdHoc::on_attach` method.

The method probably can be implemented generically in standard Rust, but I don't think we'll get something as clean as that. I'll play with it and see what I can see.